### DISCUSSION TITLE
General Performance Best Practices

### DISCUSSION THREAD

**Discussion by julianhall** [sebastienros/jint#1856] [General]
Hi -
We've been following this project for a while (well close to a decade!).  We've implemented a few POC's in attempt to replace our internal JavaScript engine from, i believe, this one (https://github.com/PureKrome/EcmaScript.NET).  It's ancient and we need to replace our internal js engine in our .NET web app. (proprietary language believe it or not...)

I'm looking to see if anyone has some general guidance on improving performance. Here are some details:
- no need to use any .net classes
- many small scripts used for conditional evaluations in a web request
- the scripts unfortunately contain proprietary variables/syntax and are first parsed/evaluated prior to sending to the JINT engine for evaluation. (i mention this as on the main readme under performance - parsed scripts).

Thanks in advance.

Jules

**Comment by SebastianStehle** [comment]
Just two things I do:

1. Cache the scripts.
2. Use lazy initialization for complex object graphs. If you have to bring a complex context to jint you might want to have custom classes. Just initialize the properties when they are needed.

**Reply by julianhall** [reply]
Thanks - Yes re:cache - I mentioned above that the scripts are dynamically created/evaluated due to certain restrictions.  We will investigate further.

**Reply by SebastianStehle** [reply]
Sure, but you can still use some kind of LRU cache. E.g. when you use the same expression in a workflow dozens of times.

**Comment by lahma** [comment]
I'd suggest measuring your specific use cases. I'm not familiar with EcmaScript.NET, but I run our standard engine comparison benchmark like so (regex test was stuck against EcmaScript.NET so skipped it):

```c#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Order;

namespace Jint.Benchmark;

[RankColumn]
[MemoryDiagnoser]
[Orderer(SummaryOrderPolicy.FastestToSlowest)]
[GroupBenchmarksBy(BenchmarkLogicalGroupRule.ByParams)]
[HideColumns("Error", "Gen0", "Gen1", "Gen2")]
[BenchmarkCategory("EngineComparison")]
public class EngineComparisonBenchmark
{
    private static readonly Dictionary<string, Prepared<Script>> _parsedScripts = new();

    private static readonly Dictionary<string, string> _files = new()
    {
        { "array-stress", null },
        { "evaluation", null },
        { "linq-js", null },
        { "minimal", null },
        { "stopwatch", null },
        { "dromaeo-3d-cube", null },
        { "dromaeo-core-eval", null },
        { "dromaeo-object-array", null },
        //{ "dromaeo-object-regexp", null },
        { "dromaeo-object-string", null },
        { "dromaeo-string-base64", null },
    };

    private static readonly string _dromaeoHelpers = @"
        var startTest = function () { };
        var test = function (name, fn) { fn(); };
        var endTest = function () { };
        var prep = function (fn) { fn(); };";

    [GlobalSetup]
    public void Setup()
    {
        foreach (var fileName in _files.Keys.ToList())
        {
            var script = File.ReadAllText($"Scripts/{fileName}.js");
            if (fileName.Contains("dromaeo"))
            {
                script = _dromaeoHelpers + Environment.NewLine + Environment.NewLine + script;
            }
            _files[fileName] = script;
            _parsedScripts[fileName] = Engine.PrepareScript(script, strict: true);
        }
    }

    [ParamsSource(nameof(FileNames))]
    public string FileName { get; set; }

    public IEnumerable<string> FileNames()
    {
        return _files.Select(entry => entry.Key);
    }

    [Benchmark]
   public void Jint()
   {
       var engine = new Engine(static options => options.Strict());
       engine.Execute(_files[FileName]);
   }

   [Benchmark]
   public void Jint_ParsedScript()
   {
       var engine = new Engine(static options => options.Strict());
       engine.Execute(_parsedScripts[FileName]);
   }

    [Benchmark(Baseline = true)]
    public void EcmaScriptNet()
    {
        var engine = EcmaScript.NET.Context.Enter();
        var scope = engine.InitStandardObjects();
        engine.EvaluateString(scope, _files[FileName], FileName, 0, null);
    }
}
```


And got results:

```
BenchmarkDotNet v0.13.12, Windows 11 (10.0.26120.461)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK 8.0.204
  [Host]     : .NET 8.0.4 (8.0.424.16909), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.4 (8.0.424.16909), X64 RyuJIT AVX2
```

| Method            | FileName             | Mean             | StdDev          | Median           | Ratio | RatioSD | Rank | Allocated      | Alloc Ratio |
|------------------ |--------------------- |-----------------:|----------------:|-----------------:|------:|--------:|-----:|---------------:|------------:|
| Jint              | array-stress         |     6,797.228 us |      52.2542 us |     6,795.374 us |  0.32 |    0.01 |    1 |     6753.54 KB |        0.70 |
| Jint_ParsedScript | array-stress         |     7,195.196 us |      98.0526 us |     7,161.510 us |  0.34 |    0.01 |    2 |     6719.28 KB |        0.70 |
| EcmaScriptNet     | array-stress         |    21,143.619 us |     423.8689 us |    20,988.959 us |  1.00 |    0.00 |    3 |     9657.31 KB |        1.00 |
|                   |                      |                  |                 |                  |       |         |      |                |             |
| EcmaScriptNet     | dromaeo-3d-cube      |    10,030.746 us |     128.1168 us |     9,979.109 us |  1.00 |    0.00 |    1 |     4492.98 KB |        1.00 |
| Jint_ParsedScript | dromaeo-3d-cube      |    16,634.849 us |     306.3860 us |    16,567.242 us |  1.66 |    0.02 |    2 |     5880.61 KB |        1.31 |
| Jint              | dromaeo-3d-cube      |    17,368.136 us |     272.3706 us |    17,286.412 us |  1.73 |    0.04 |    3 |     6229.64 KB |        1.39 |
|                   |                      |                  |                 |                  |       |         |      |                |             |
| Jint              | dromaeo-core-eval    |     3,305.923 us |      30.4869 us |     3,301.874 us |  0.61 |    0.02 |    1 |      352.52 KB |       0.011 |
| Jint_ParsedScript | dromaeo-core-eval    |     3,943.948 us |      31.5856 us |     3,943.395 us |  0.73 |    0.02 |    2 |      329.37 KB |       0.010 |
| EcmaScriptNet     | dromaeo-core-eval    |     5,390.015 us |     116.1087 us |     5,353.594 us |  1.00 |    0.00 |    3 |    33456.23 KB |       1.000 |
|                   |                      |                  |                 |                  |       |         |      |                |             |
| Jint_ParsedScript | dromaeo-object-array |    43,984.115 us |     111.5653 us |    43,954.083 us |  0.17 |    0.00 |    1 |    96244.74 KB |        0.98 |
| Jint              | dromaeo-object-array |    44,476.857 us |   1,214.0549 us |    44,244.533 us |  0.17 |    0.00 |    1 |    96299.97 KB |        0.98 |
| EcmaScriptNet     | dromaeo-object-array |   259,423.025 us |   2,447.9582 us |   259,071.875 us |  1.00 |    0.00 |    2 |    98659.68 KB |        1.00 |
|                   |                      |                  |                 |                  |       |         |      |                |             |
| Jint_ParsedScript | droma(...)tring [21] |   245,650.433 us |   9,148.0762 us |   247,704.600 us |  0.05 |    0.00 |    1 |  1315788.11 KB |        0.06 |
| Jint              | droma(...)tring [21] |   249,525.543 us |  11,622.0893 us |   247,168.700 us |  0.05 |    0.00 |    1 |   1315710.4 KB |        0.06 |
| EcmaScriptNet     | droma(...)tring [21] | 5,276,508.742 us | 400,919.8263 us | 5,079,163.050 us |  1.00 |    0.00 |    2 | 20558703.97 KB |        1.00 |
|                   |                      |                  |                 |                  |       |         |      |                |             |
| Jint_ParsedScript | droma(...)ase64 [21] |    37,712.683 us |     133.3056 us |    37,721.654 us |  0.35 |    0.03 |    1 |     2329.59 KB |       0.003 |
| Jint              | droma(...)ase64 [21] |    46,404.408 us |   6,406.1056 us |    42,627.542 us |  0.46 |    0.09 |    2 |     2438.29 KB |       0.003 |
| EcmaScriptNet     | droma(...)ase64 [21] |   102,806.507 us |  13,896.0370 us |    95,754.767 us |  1.00 |    0.00 |    3 |   760793.07 KB |       1.000 |
|                   |                      |                  |                 |                  |       |         |      |                |             |
| Jint_ParsedScript | evaluation           |         7.692 us |       0.0624 us |         7.663 us |  0.23 |    0.00 |    1 |       24.41 KB |        0.50 |
| Jint              | evaluation           |        21.076 us |       0.2040 us |        21.086 us |  0.64 |    0.01 |    2 |          36 KB |        0.74 |
| EcmaScriptNet     | evaluation           |        33.150 us |       0.3757 us |        33.179 us |  1.00 |    0.00 |    3 |       48.47 KB |        1.00 |
|                   |                      |                  |                 |                  |       |         |      |                |             |
| Jint_ParsedScript | linq-js              |        83.028 us |       2.0959 us |        83.041 us |  0.03 |    0.00 |    1 |      193.38 KB |        0.05 |
| Jint              | linq-js              |     1,589.961 us |      32.2168 us |     1,584.726 us |  0.66 |    0.04 |    2 |     1309.87 KB |        0.37 |
| EcmaScriptNet     | linq-js              |     2,435.237 us |     151.4443 us |     2,365.885 us |  1.00 |    0.00 |    3 |     3563.45 KB |        1.00 |
|                   |                      |                  |                 |                  |       |         |      |                |             |
| Jint_ParsedScript | minimal              |         2.094 us |       0.0710 us |         2.097 us |  0.11 |    0.00 |    1 |       13.55 KB |        0.42 |
| Jint              | minimal              |         3.519 us |       0.0705 us |         3.499 us |  0.20 |    0.00 |    2 |       15.59 KB |        0.49 |
| EcmaScriptNet     | minimal              |        17.875 us |       0.1612 us |        17.849 us |  1.00 |    0.00 |    3 |          32 KB |        1.00 |
|                   |                      |                  |                 |                  |       |         |      |                |             |
| Jint              | stopwatch            |   280,852.584 us |   4,853.7762 us |   281,805.475 us |  0.70 |    0.01 |    1 |    53038.07 KB |        0.39 |
| Jint_ParsedScript | stopwatch            |   295,099.803 us |   5,908.0872 us |   296,609.300 us |  0.73 |    0.02 |    2 |    53001.66 KB |        0.39 |
| EcmaScriptNet     | stopwatch            |   403,260.125 us |   2,264.8376 us |   402,814.700 us |  1.00 |    0.00 |    3 |   136993.96 KB |        1.00 |

**Reply by julianhall** [reply]
Marko - thank you very much for the quick analysis and response!

Looking at your results - Jint should obviously be the clear winner here.  Hmm.

I will re-run your tests with the version of the engine we have and see what's up!
